OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs

```diff
@@ -17,15 +17,23 @@ namespace OpenRA.Mods.Common.Traits
 {
 	public static class BuildingUtils
 	{
-		public static bool IsCellBuildable(this World world, CPos cell, BuildingInfo bi, Actor toIgnore = null)
+		public static bool IsCellBuildable(this World world, CPos cell, ActorInfo ai, BuildingInfo bi, Actor toIgnore = null)
 		{
 			if (!world.Map.Contains(cell))
 				return false;
 
 			if (world.WorldActor.Trait<BuildingInfluence>().GetBuildingAt(cell) != null)
-				return false;
-
-			if (!bi.AllowInvalidPlacement && world.ActorMap.GetActorsAt(cell).Any(a => a != toIgnore))
+			{
+				if (ai == null)
+					return false;
+
+				var replacementInfo = ai.TraitInfoOrDefault<ReplacementInfo>();
+				if (replacementInfo == null)
+					return false;
+				if (!IsBuildingAtCellReplaceable(world, cell, replacementInfo))
```

Comment:
Can you please store the result of the `world.WorldActor.Trait<BuildingInfluence>().GetBuildingAt(cell)` above in a variable and then query the `Replaceable` trait inline here?  `IsBuildingAtCellReplaceable` is only used from here, and most of its code is repeating queries that we already have the result for!